internal/graph/lookupresources.go

```diff
@@ -44,105 +45,113 @@ func (cl *CursoredLookupResources) LookupResources(
-							},
-						},
-						Metadata: combineResponseMetadata(result.Metadata, checkMeta),
-					})
-					continue
-
-				case v1.ResourceCheckResult_CAVEATED_MEMBER:
-					stream.Publish(&v1.DispatchLookupResourcesResponse{
-						ResolvedResources: []*v1.ResolvedResource{
-							{
-								ResourceId:             reachableResource.ResourceId,
-								Permissionship:         v1.ResolvedResource_CONDITIONALLY_HAS_PERMISSION,
-								MissingRequiredContext: checkResult.MissingExprFields,
-							},
-						},
-						Metadata: combineResponseMetadata(result.Metadata, checkMeta),
-					})
-					continue
-
-				case v1.ResourceCheckResult_NOT_MEMBER:
-					// Skip.
-					continue
-
-				default:
-					return spiceerrors.MustBugf("unknown check result status for reachable resources")
+					[]string{reachableResource.ResourceId},
+				)
+				if err != nil {
+					return err
 				}
 
+				checkResult = results[reachableResource.ResourceId]
+				metadata = combineResponseMetadata(resultsMeta, metadata)
+			}
+
+			// Publish the resource, if applicable.
+			permissionship := v1.ResolvedResource_UNKNOWN
+			var missingFields []string
+
+			switch {
+			case reachableResource.ResultStatus == v1.ReachableResource_HAS_PERMISSION:
+				permissionship = v1.ResolvedResource_HAS_PERMISSION
+
+			case checkResult == nil || checkResult.Membership == v1.ResourceCheckResult_NOT_MEMBER:
+				return nil
+
+			case checkResult != nil && checkResult.Membership == v1.ResourceCheckResult_MEMBER:
+				permissionship = v1.ResolvedResource_HAS_PERMISSION
+
+			case checkResult != nil && checkResult.Membership == v1.ResourceCheckResult_CAVEATED_MEMBER:
+				permissionship = v1.ResolvedResource_CONDITIONALLY_HAS_PERMISSION
+				missingFields = checkResult.MissingExprFields
+
 			default:
-				return spiceerrors.MustBugf("unknown result status for reachable resources")
+				return spiceerrors.MustBugf("unknown check result status for reachable resources")
 			}
+
+			okay, done := limits.prepareForPublishing()
+			defer done()
+
+			if !okay {
+				return nil
+			}
+
+			return stream.Publish(&v1.DispatchLookupResourcesResponse{
+				ResolvedResource: &v1.ResolvedResource{
+					ResourceId:             reachableResource.ResourceId,
+					Permissionship:         permissionship,
+					MissingRequiredContext: missingFields,
+				},
+				Metadata:            metadata,
+				AfterResponseCursor: result.AfterResponseCursor,
+			})
+		})
+
+		err := cl.r.DispatchReachableResources(&v1.DispatchReachableResourcesRequest{
+			ResourceRelation: req.ObjectRelation,
+			SubjectRelation: &core.RelationReference{
+				Namespace: req.Subject.Namespace,
+				Relation:  req.Subject.Relation,
+			},
+			SubjectIds:     []string{req.Subject.ObjectId},
+			Metadata:       req.Metadata,
+			OptionalCursor: reachableResourcesCursor,
+			OptionalLimit:  reachableResourcesLimit,
+		}, rrStream)
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return nil
+			}
```

Comment:
Because we always cancel reachable resources when LR meets its limit; propagating the context canceled from reachable resources means that LR will never not return an error, unless we happen to be at the exact border of the limit.

Reply:
ah, got it 👍🏻